src/Jellyfin.MediaEncoding.Hls/ScheduledTasks/KeyframeExtractionScheduledTask.cs

```diff
@@ -0,0 +1,100 @@
+
+/// <inheritdoc />
+public class KeyframeExtractionScheduledTask : IScheduledTask
+{
+    private readonly ILocalizationManager _localizationManager;
+    private readonly ILibraryManager _libraryManager;
+    private readonly IKeyframeExtractor[] _keyframeExtractors;
+    private static readonly BaseItemKind[] _itemTypes = { BaseItemKind.Episode, BaseItemKind.Movie };
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="KeyframeExtractionScheduledTask"/> class.
+    /// </summary>
+    /// <param name="localizationManager">An instance of the <see cref="ILocalizationManager"/> interface.</param>
+    /// <param name="libraryManager">An instance of the <see cref="ILibraryManager"/> interface.</param>
+    /// <param name="keyframeExtractors">The keyframe extractors.</param>
+    public KeyframeExtractionScheduledTask(ILocalizationManager localizationManager, ILibraryManager libraryManager, IEnumerable<IKeyframeExtractor> keyframeExtractors)
+    {
+        _localizationManager = localizationManager;
+        _libraryManager = libraryManager;
+        _keyframeExtractors = keyframeExtractors.ToArray();
+    }
+
+    /// <inheritdoc />
+    public string Name => "Keyframe Extractor";
+
+    /// <inheritdoc />
+    public string Key => "KeyframeExtraction";
+
+    /// <inheritdoc />
+    public string Description => "Extracts keyframes from video files to create more precise HLS playlists";
+
+    /// <inheritdoc />
+    public string Category => _localizationManager.GetLocalizedString("TasksLibraryCategory");
+
+    /// <inheritdoc />
+    public Task Execute(CancellationToken cancellationToken, IProgress<double> progress)
+    {
+        var query = new InternalItemsQuery
+        {
+            MediaTypes = new[] { MediaType.Video },
+            IsVirtualItem = false,
+            IncludeItemTypes = _itemTypes,
+            DtoOptions = new DtoOptions(true),
+            SourceTypes = new[] { SourceType.Library },
+            Recursive = true
+        };
+
+        var videos = _libraryManager.GetItemList(query);
+        var numberOfVideos = videos.Count;
+        var numComplete = 0;
+
+        // TODO parallelize with Parallel.ForEach?
+        for (var i = 0; i < numberOfVideos; i++)
+        {
+            var video = videos[i];
+            // Only local files supported
+            if (!video.IsFileProtocol || !File.Exists(video.Path))
+            {
+                continue;
+            }
+
+            for (var j = 0; j < _keyframeExtractors.Length; j++)
+            {
+                var extractor = _keyframeExtractors[j];
+                // The cache decorator will make sure to save them in the data dir
+                if (extractor.TryExtractKeyframes(video.Path, out _))
+                {
+                    break;
+                }
+            }
+
+            // Update progress
+            numComplete++;
+            double percent = (double)numComplete / numberOfVideos;
```

Comment:
Probably ok, but this will never be 100% if there are any remote videos

Reply:
Given our brilliant progress bars in general, I don't think it would be too much of an issue right?